src/Forms/Form.php

```diff
@@ -317,6 +322,58 @@ public function __construct(
         $this->setupDefaultClasses();
     }
 
+    /**
+     * Make the form require sudo mode, which will make the form readonly and add a sudo mode password field
+     * unless the current user previously activated sudo mode.
+     *
+     * Note that if the parent request handler for this form isn't LeftAndMain or GridFieldDetailForm_ItemRequest,
+     * sudo mode will not be required by this form.
+     */
+    public function requireSudoMode(): void
+    {
+        // Check that the current request handler for the form is one that's used
+        // in an admin context where sudo mode makes sense
+        $classes = [
+            LeftAndMain::class,
+            GridFieldDetailForm_ItemRequest::class,
+        ];
+        $enableSudoMode = false;
+        foreach ($classes as $class) {
+            if (is_a($this->getController(), $class)) {
+                $enableSudoMode = true;
+                break;
+            }
+        }
+        if (!$enableSudoMode) {
+            return;
+        }
+        // Check if sudo mode is currently enabled
+        $service = Injector::inst()->get(SudoModeServiceInterface::class);
+        $session = $this->getRequest()->getSession();
+        if ($service->check($session)) {
+            return;
+        }
+        // If sudo mode is not active, make the form readonly and add a sudo mode password field
+        // Modify the GridFieldViewButton on any GridFields so that it doesn't suffix the view URL with 'view'
+        // This allows us to gracefully reload a form in readonly mode when sudo mode is activated
+        foreach ($this->Fields()->flattenFields() as $field) {
+            if (!is_a($field, GridField::class)) {
+                continue;
+            }
+            $config = $field->getConfig();
+            $config->removeComponentsByType(GridFieldViewButton::class);
+            $viewButton = GridFieldViewButton::create();
+            $viewButton->setSuffixViewToUrl(false);
+            $config->addComponent($viewButton);
+        }
```

Comment:
This is fine for if the _whole form_ is readonly, but doesn't work for this scenario:

```php
    public function getCMSFields()
    {
        $fields = parent::getCMSFields();
        $fields->addFieldToTab('Root.Main', GridField::create('Members', 'Members', Member::get(), GridFieldConfig_RecordEditor::create()));
        return $fields;
    }
```

That gridfield is still readonly with a normal unaltered view button component.

I think this code actually doesn't need to be in `Form` at all - just in `GridField::Fieldholder()`

Reply:
Yeah much nicer in `GridField::Fieldholder()`, will do there

Reply:
Looks like you forgot to push